File: server/Src/DabaebmintonServerCore/AckRange.cpp
#include "ServerPCH.h"
#include "AckRange.h"

#include "InputMemoryBitStream.h"

void AckRange::Read(InputMemoryBitStream& inStream)
{
	inStream.Read(_start);
	bool hasCount{};
	inStream.Read(hasCount);

	if (hasCount)	// hasCount == false: count 1개, 
					// hasCount == true: count 2개 이상
	{
		inStream.Read(_count);
	}
	else
	{
		_count = 1;	// default.
	}
}

File: client/DabaebMinton/Assets/Scripts/Network/AckRange.cs
using System;
using Unity.VisualScripting;
using UnityEngine;
using PacketSequenceNumber = System.UInt16;

public class AckRange
{
    public AckRange(PacketSequenceNumber inStart)
    {
        _start = inStart;
        _count = 1;
    }

    public bool ExtendIfShould(PacketSequenceNumber inSequenceNumber)
    {
        if (inSequenceNumber == _start + _count)
        {
            ++_count;
            return true;
        }

        return false;
    }

    public void Write(OutputMemoryBitStream outStream)
    {
        outStream.Write(_start);
        bool hasCount = _count > 1;
        outStream.Write(hasCount);
        if (hasCount)
        {
            outStream.Write(_count);
        }
    }

    private PacketSequenceNumber _start, _count;
}
